File: apps/api/app/schemas/calls.py
"""Pydantic schemas for extracted call sites."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class SymbolCallRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    snapshot_id: UUID
    source_file_id: UUID
    path: str
    caller_symbol_id: UUID | None
    caller_qualified_name: str | None
    raw_callee: str
    qualified_expression: str
    line: int
    candidate_qualified_name: str | None
    confidence: str
    language: str
    created_at: datetime


class SymbolCallListResponse(BaseModel):
    repository_id: UUID
    snapshot_id: UUID | None
    total: int
    limit: int
    offset: int
    calls: list[SymbolCallRead]


class SymbolNeighborsResponse(BaseModel):
    """Callers or callees for one symbol (Week 4 Day 7)."""

    repository_id: UUID
    snapshot_id: UUID
    symbol_id: UUID
    symbol_qualified_name: str
    direction: str  # callers | callees
    total: int
    calls: list[SymbolCallRead]

File: apps/web/src/lib/calls.ts
export interface SymbolCallItem {
  id: string;
  snapshot_id: string;
  source_file_id: string;
  path: string;
  caller_symbol_id: string | null;
  caller_qualified_name: string | null;
  raw_callee: string;
  qualified_expression: string;
  line: number;
  candidate_qualified_name: string | null;
  confidence: string;
  language: string;
  created_at: string;
}

export interface SymbolCallListResponse {
  repository_id: string;
  snapshot_id: string | null;
  total: number;
  limit: number;
  offset: number;
  calls: SymbolCallItem[];
}

export interface SymbolNeighborsResponse {
  repository_id: string;
  snapshot_id: string;
  symbol_id: string;
  symbol_qualified_name: string;
  direction: "callers" | "callees" | string;
  total: number;
  calls: SymbolCallItem[];
}

export function confidenceLabel(confidence: string): string {
  switch (confidence) {
    case "resolved":
      return "Resolved";
    case "ambiguous":
      return "Ambiguous";
    case "unresolved":
      return "Unresolved";
    default:
      return confidence;
  }
}
